File: 0785-is-graph-bipartite/0785-is-graph-bipartite.cpp
class Solution {
public:
    bool dfs(int node, int col, int vis[], vector<vector<int>>& graph) {
        vis[node] = col;
        for (auto it : graph[node]) {
            if (vis[it] == -1) {
                if (dfs(it, !col, vis, graph) == false) {
                    return false;
                }
            } else if (vis[it] == col) {
                return false;
            }
        }
        return true;
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        int vis[n];
        for (int i = 0; i < n; i++) {
            vis[i] = -1;
        }
        for (int i = 0; i < n; i++) {
            if (vis[i] == -1) {
                if (dfs(i, 0, vis, graph) == false) {
                    return false;
                }
            }
        }
        return true;
    }
};

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.java
class Solution {
    public boolean bipart(int st,int[] color,int[][] graph){
        Queue<Integer>q=new LinkedList<>();
        q.offer(st);
        color[st]=0;

        while(!q.isEmpty()){
            int node=q.poll();

            for(int nxt:graph[node]){
                if(color[nxt]==-1){
                    color[nxt]=1-color[node];
                    q.offer(nxt);
                }else if(color[nxt]==color[node]){
                    return false;
                }
            }
        }
        return true;
    }
    public boolean isBipartite(int[][] graph) {
        int n=graph.length;
        int[] color=new int[n];

        Arrays.fill(color,-1);

        for(int i=0;i<n;i++){
            if(color[i]==-1){
                if(!bipart(i,color,graph)){
                    return false;
                }
            }
        }
        return true;
    }
}